lib/elixir/lib/module/types/descr.ex

```diff
@@ -839,6 +876,549 @@ defmodule Module.Types.Descr do
+  # handling of unions, intersections, and negations.
+
+  ### Key concepts:
+
+  # * BDD structure: A tree with function nodes and 0/1 leaves. Paths to leaf 1
+  #   represent valid function types. Nodes are positive when following a left
+  #   branch (e.g. (int, float) -> bool) and negative otherwise.
+
+  # * Function variance:
+  #   - Contravariance in arguments: If s <: t, then (t → r) <: (s → r)
+  #   - Covariance in returns: If s <: t, then (u → s) <: (u → t)
+
+  # * Representation:
+  #   - fun(): Top function type (leaf 1)
+  #   - Function literals: {tag, [t1, ..., tn], t} where [t1, ..., tn] are argument types and t is return type
+  #     tag is either `:weak` or `:strong`
+  #     TODO: implement `:strong`
+  #   - Normalized form for function applications: {domain, arrows, arity} is produced by `fun_normalize/1`
+
+  # * Examples:
+  #   - fun([integer()], atom()): A function from integer to atom
+  #   - intersection(fun([integer()], atom()), fun([float()], boolean())): A function handling both signatures
+
+  # Note: Function domains are expressed as tuple types. We use separate representations rather than
+  # unary functions with tuple domains to handle special cases like representing functions of a
+  # specific arity (e.g., (none,none->term) for arity 2).
+
+  defp fun_new(inputs, output), do: {{:weak, inputs, output}, 1, 0}
+
+  @doc """
+  Creates a function type from a list of inputs and an output where the inputs and/or output may be dynamic.
+
+  For function (t → s) with dynamic components:
+  - Static part:  (up(t) → down(s))
+  - Dynamic part: dynamic(down(t) → up(s))
+
+  When handling dynamic types:
+  - `up(t)` extracts the upper bound (most general type) of a gradual type.
+    For `dynamic(integer())`, it is `integer()`.
+  - `down(t)` extracts the lower bound (most specific type) of a gradual type.
+  """
+  def fun_descr(args, output) when is_list(args) do
+    dynamic_arguments? = are_arguments_dynamic?(args)
+    dynamic_output? = match?(%{dynamic: _}, output)
+
+    if dynamic_arguments? or dynamic_output? do
+      input_static = if dynamic_arguments?, do: materialize_arguments(args, :up), else: args
+      input_dynamic = if dynamic_arguments?, do: materialize_arguments(args, :down), else: args
+
+      output_static = if dynamic_output?, do: down(output), else: output
+      output_dynamic = if dynamic_output?, do: up(output), else: output
+
+      %{
+        fun: fun_new(input_static, output_static),
+        dynamic: %{fun: fun_new(input_dynamic, output_dynamic)}
+      }
+    else
+      # No dynamic components, use standard function type
+      %{fun: fun_new(args, output)}
+    end
+  end
+
+  # Gets the upper bound of a gradual type.
+  defp up(%{dynamic: dynamic}), do: dynamic
+  defp up(static), do: static
+
+  # Gets the lower bound of a gradual type.
+  defp down(:term), do: :term
+  defp down(type), do: Map.delete(type, :dynamic)
```

Comment:
Let's either call it `upper_bound` or `dynamic_upper`, if we plan to use it in other places, or we  call it `fun_up` or `fun_down`. :)